sparse/impl/KokkosSparse_par_ilut_symbolic_impl.hpp

```diff
@@ -0,0 +1,176 @@
+                   LEntriesType& L_entries_d, LValuesType& L_values_d,
+                   URowMapType& U_row_map_d, UEntriesType& U_entries_d,
+                   UValuesType& U_values_d) {
+  using execution_space = typename ARowMapType::execution_space;
+  using policy_type     = Kokkos::TeamPolicy<execution_space>;
+  using member_type     = typename policy_type::member_type;
+  using size_type       = typename IlutHandle::size_type;
+  using scalar_t        = typename AValuesType::non_const_value_type;
+  using index_t         = typename AEntriesType::non_const_value_type;
+  using RangePolicy     = typename IlutHandle::RangePolicy;
+
+  const size_type nrows = thandle.get_nrows();
+
+  const auto policy = thandle.get_default_team_policy();
+
+  // Sizing
+  Kokkos::parallel_for(
+      "symbolic sizing", policy, KOKKOS_LAMBDA(const member_type& team) {
+        const auto row_idx = team.league_rank();
+
+        const auto row_nnz_begin = A_row_map_d(row_idx);
+        const auto row_nnz_end   = A_row_map_d(row_idx + 1);
+
+        size_type nnzsL_temp = 0, nnzsU_temp = 0;
+        // Multi-reductions are not supported at the TeamThread level
+        Kokkos::parallel_reduce(
+            Kokkos::TeamThreadRange(team, row_nnz_begin, row_nnz_end),
+            [&](const size_type nnz, size_type& nnzsL_inner) {
+              const auto col_idx = A_entries_d(nnz);
+              nnzsL_inner += col_idx < row_idx;
+            },
+            nnzsL_temp);
+
+        Kokkos::parallel_reduce(
+            Kokkos::TeamThreadRange(team, row_nnz_begin, row_nnz_end),
+            [&](const size_type nnz, size_type& nnzsU_inner) {
+              const auto col_idx = A_entries_d(nnz);
+              nnzsU_inner += col_idx > row_idx;
+            },
+            nnzsU_temp);
+
+        team.team_barrier();
+
+        Kokkos::single(Kokkos::PerTeam(team), [&]() {
+          L_row_map_d(row_idx) = nnzsL_temp + 1;
+          U_row_map_d(row_idx) = nnzsU_temp + 1;
+        });
+      });
+
+  const size_type nnzsL = prefix_sum<IlutHandle>(L_row_map_d);
+  const size_type nnzsU = prefix_sum<IlutHandle>(U_row_map_d);
+
+  thandle.set_nnzL(nnzsL);
+  thandle.set_nnzU(nnzsU);
+
+  // Now set actual L/U values
+
+  Kokkos::parallel_for(
+      "symbolic values",
+      RangePolicy(0, nrows),  // No team level parallelism in this alg
+      KOKKOS_LAMBDA(const index_t& row_idx) {
+        const auto row_nnz_begin = A_row_map_d(row_idx);
+        const auto row_nnz_end   = A_row_map_d(row_idx + 1);
+
+        size_type current_index_l = L_row_map_d(row_idx);
+        size_type current_index_u =
+            U_row_map_d(row_idx) + 1;  // we treat the diagonal separately
+
+        // if there is no diagonal value, set it to 1 by default
+        scalar_t diag = 1.;
+
+        for (size_type row_nnz = row_nnz_begin; row_nnz < row_nnz_end;
+             ++row_nnz) {
+          const auto val     = A_values_d(row_nnz);
```

Comment:
The symbolic phase shouldn't depend on A's values. I think this second kernel that fills in the entries/values of L/U should be in numeric.

Ideally the pattern (similar to spadd/spgemm) would go:
- symbolic figures out nnzsL and nnzsU and saves those in the handle
- the user gets those back and allocates L_entries, L_values, U_entries, U_values
- then numeric gets called and fills those in.

This has some benefits besides consistency:
- none of the input views you pass through the internal functions (``PAR_ILUT_SYMBOLIC::par_ilut_symbolic`` and ``PAR_ILUT_NUMERIC::par_ilut_numeric``) can be passed by constant reference, so you can avoid doing the reinterpret_casts to unified view types in KokkosSparse_par_ilut.hpp. I'm not sure that reinterpret-casting a managed view to unmanaged is safe in general.
- the result of the symbolic phase can be reused if A's values change, as long as its structure stays the same. This is the really the purpose of the symbolic/numeric separation.
- the symbolic phase functions can have fewer template and regular parameters - nothing involving scalars/values.